Sources/Dependencies/DependencyValues.swift

```diff
@@ -13,23 +15,40 @@
 /// @Dependency(\.date) var date
 /// ```
 public struct DependencyValues: Sendable {
+  public enum Environment {
+    case live, preview, test
+  }
+
   @TaskLocal public static var current = Self()
 
   private var storage: [ObjectIdentifier: AnySendable] = [:]
 
   public init() {}
 
-  public init(isTesting: Bool) {
-    self.isTesting = isTesting
-  }
-
   public subscript<Key: DependencyKey>(key: Key.Type) -> Key.Value where Key.Value: Sendable {
     get {
       guard let dependency = self.storage[ObjectIdentifier(key)]?.base as? Key.Value
       else {
-        let isTesting = self.storage[ObjectIdentifier(IsTestingKey.self)]?.base as? Bool ?? false
-        guard !isTesting else { return Key.testValue }
-        return _liveValue(Key.self) as? Key.Value ?? Key.testValue
+
+        let mode =
+          self.storage[ObjectIdentifier(EnvironmentKey.self)]?.base as? Environment
+          ?? {
+            #if DEBUG
+              if ProcessInfo.processInfo.environment["XCODE_RUNNING_FOR_PREVIEWS"] == "1" {
```

Comment:
I didn't check with this exact setup, but I had warnings while polling this dictionary in async contexts. If so, the value could maybe be stored in a top-level constant property as a workaround.

Reply:
@tgrapperon Ah interesting, yeah would be a bummer to spew warnings into logs unnecessarily. Just pushed a commit to compute this value a single time.

Reply:
It was build-time warnings in my case, because I was touching a mutable shared state. But I was also using it inside by CoreData abstraction to force some system entities to be defined synchronously when previewing. It is possible that in this case, it could be OK. As I'm getting it, this subscript can be hit from any async context, so I guess should have complained too with warnings in place, but I didn't check.

Reply:
ah ok, well moving it out to a helper makes the code easier to read anyway, so win-win!